pkg/webhook/cluster/konnectoragent_legacy_validator.go

```diff
@@ -0,0 +1,239 @@
+// Copyright 2025 Nutanix. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package cluster
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/release"
+	v1 "k8s.io/api/admission/v1"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/rest"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/api/v1alpha1"
+	apivariables "github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/api/variables"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/variables"
+)
+
+const (
+	// legacyHelmChartName is the chart name of the old helm release
+	// that needs to be detected during upgrades.
+	legacyHelmChartName = "nutanix-k8s-agent"
+)
+
+type konnectorAgentLegacyValidator struct {
+	client  ctrlclient.Client
+	decoder admission.Decoder
+}
+
+func NewKonnectorAgentLegacyValidator(
+	client ctrlclient.Client, decoder admission.Decoder,
+) *konnectorAgentLegacyValidator {
+	return &konnectorAgentLegacyValidator{
+		client:  client,
+		decoder: decoder,
+	}
+}
+
+func (k *konnectorAgentLegacyValidator) Validator() admission.HandlerFunc {
+	return k.validate
+}
+
+func (k *konnectorAgentLegacyValidator) validate(
+	ctx context.Context,
+	req admission.Request,
+) admission.Response {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Only validate on UPDATE operations
+	// Skip CREATE and DELETE operations
+	if req.Operation != v1.Update {
+		log.Info("Skipping validation for non-UPDATE operations", "operation", req.Operation)
```

Comment:
This is pretty chatty and doesn't really provide much value in a running controller, the integration tests can very these conditions are handled correctly.
```suggestion
```